### PR TITLE
Add failpoints for network sending (for forge tests)

### PR DESCRIPTION
**Pull request by igor-aptos**
Add network::send::any failpoint, which individually decides (per message, even for broadcast), whether to send it or not.

(otherwise, if broadcast is decided in bulk, small failure %, can make nobody receive a broadcast, which is not realistic)

### CODE REVIEW

File: network/src/peer_manager/senders.rs [hunk 1]
```diff
@@ -100,6 +110,10 @@ impl PeerManagerRequestSender {
             res_tx,
             timeout,
         };
+        fail_point!("network::send::any", |_| {
+            std::thread::sleep(timeout);
```
Comment (JoshLind) [inline network/src/peer_manager/senders.rs:114]: You probably want tokio::time::sleep, no? Otherwise this will block the entire OS thread?

File: network/src/peer_manager/senders.rs [hunk 2]
```diff
@@ -47,6 +48,10 @@ impl PeerManagerRequestSender {
         protocol_id: ProtocolId,
         mdata: Bytes,
     ) -> Result<(), PeerManagerError> {
+        fail_point!("network::send::any", |_| {
```
Comment (JoshLind) [inline network/src/peer_manager/senders.rs:51]: Is your goal to add this only for consensus? Or for all components? Right now, this will affect everything, right?
Also, why did you add this to the `PeerManagerRequestSender` and not at the network interface level? e.g., https://github.com/aptos-labs/aptos-core/blob/f005fa18b1095aad2ba4f128e5da88c89e4d164b/network/src/application/interface.rs#L166. 

If you want to add it to everything, the interface feels like a cleaner place?

Comment (danielxiangzl) [reply network/src/peer_manager/senders.rs:51]: +1. But it looks like there is no easy way to add fail points only for consensus, without changing the existing networking?

File: network/src/peer_manager/senders.rs (hunk 1 above)
Comment (zekun000) [reply network/src/peer_manager/senders.rs:114]: it's probably not easy to plug in async function here, but not sure

File: network/src/peer_manager/senders.rs (hunk 2 above)
Comment (zekun000) [reply network/src/peer_manager/senders.rs:51]: we can decide based on protocol id but probably not ideal

Comment (igor-aptos) [reply network/src/peer_manager/senders.rs:51]: I wanted to have an unreliable network - for everything, not just for consensus, so it is a realistic scenario. If RPC being delayed, blocks other RPCs from other protocols, and causes issues, we do want that.

for consensus only, we have consensus::send::any already


For why I placed it there, I wanted to have a separate decision for each destination peer, which is simpler to do in PeerManagerRequestSender, because network interface level does send_to_many in a single call

**Review by JoshLind — APPROVED** [review]
LGTM. But I think the sleep should be `tokio::time::sleep`?

File: network/src/peer_manager/senders.rs (hunk 1 above)
Comment (igor-aptos) [reply network/src/peer_manager/senders.rs:114]: @JoshLind  - don't know how to have async inside of failpoint lambda?

File: network/src/peer_manager/senders.rs (hunk 2 above)
Comment (sitalkedia) [reply network/src/peer_manager/senders.rs:51]: This looks reasonable to me but wondering if we should use this for realistic workload scenario or use chaos monkey's network packet drop support. Later seems more realistic to me.

File: testsuite/testcases/src/modifiers.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        }
+        Ok::<(), anyhow::Error>(())
+    })
+}
+
+/// Config for additing variable processing overhead/delay into
+/// execution, to make different nodes have different processing speed.
+pub struct ExecutionDelayConfig {
+    /// Fraction (0.0 - 1.0) of nodes on which any delay will be introduced
+    pub inject_delay_node_fraction: f64,
```
Comment (sitalkedia) [inline testsuite/testcases/src/modifiers.rs:78]: nit: better to make it percentage instead of fraction to be consistent

File: testsuite/testcases/src/modifiers.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    /// For nodes with delay, what percentage (0-100) of transaction will be delayed.
+    /// (this is needed because delay that can be introduced is integer number of ms)
+    /// Different node speed come from this setting, each node is selected a number
+    /// between 1 and given max.
+    pub inject_delay_max_transaction_percentage: u32,
```
Comment (sitalkedia) [inline testsuite/testcases/src/modifiers.rs:83]: Can you clarify why is this needed? Why not varying `inject_delay_per_transaction_ms` per node is not sufficient?

File: testsuite/testcases/src/modifiers.rs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    /// Fixed busy-loop delay applied to each transaction that is delayed,
+    /// before it is executed.
+    pub inject_delay_per_transaction_ms: u32,
```
Comment (sitalkedia) [inline testsuite/testcases/src/modifiers.rs:86]: Does this need to be lower granularity than `ms` - 1 ms delay per txn is very high..

File: network/src/peer_manager/senders.rs (hunk 1 above)
Comment (JoshLind) [reply network/src/peer_manager/senders.rs:114]: Aah, gotcha. Yeah, not sure if failpoints supports that. In which case, ignore me 😄

**Review by davidiw — APPROVED** [review]

**Review by alokeutpal — APPROVED** [review]

**Comment by davidiw** [comment]
@gelash and team, it might be worth looking at this PR as it seems to trigger the following:

```[aptos_vm] Transaction breaking invariant violation. txn: Ok([90, 120, 118, 196, 241, 97, 45, 148, 77, 27, 231, 239, 120, 137, 74, 39, 247, 239, 223, 157, 160, 20, 194, 255, 171, 80, 251, 184, 53, 251, 4, 147, 22, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 4, 99, 111, 105, 110, 8, 116, 114, 97, 110, 115, 102, 101, 114, 1, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 10, 97, 112, 116, 111, 115, 95, 99, 111, 105, 110, 9, 65, 112, 116, 111, 115, 67, 111, 105, 110, 0, 2, 32, 2, 91, 61, 229, 231, 160, 12, 101, 58, 43, 216, 182, 67, 254, 191, 211, 20, 10, 72, 144, 30, 164, 88, 182, 85, 198, 34, 203, 205, 57, 140, 197, 8, 1, 0, 0, 0, 0, 0, 0, 0, 128, 132, 30, 0, 0, 0, 0, 0, 100, 0, 0, 0, 0, 0, 0, 0, 209, 248, 160, 100, 0, 0, 0, 0, 4, 0, 32, 85, 192, 214, 194, 59, 148, 20, 126, 199, 128, 102, 216, 101, 129, 85, 169, 220, 148, 40, 65, 46, 37, 206, 201, 171, 236, 98, 158, 221, 72, 159, 178, 64, 6, 247, 151, 45, 178, 93, 15, 232, 58, 2, 51, 81, 27, 179, 164, 172, 69, 47, 234, 246, 13, 120, 28, 21, 220, 10, 231, 219, 149, 60, 78, 6, 179, 53, 214, 44, 25, 126, 103, 210, 9, 12, 174, 42, 74, 139, 53, 210, 96, 197, 116, 134, 145, 82, 209, 158, 52, 88, 76, 94, 251, 93, 208, 9]), status: ERROR { status_code: UNEXPECTED_ERROR_FROM_KNOWN_MOVE_FUNCTION, message: None }```
